qiskit_machine_learning/neural_networks/estimator_qnn.py

```diff
@@ -0,0 +1,249 @@
+        else:
+            if weights is not None:
+                num_samples = 1
+                parameters = np.broadcast_to(weights, (num_samples, len(weights)))
+            else:
+                return None, None
+        return parameters, num_samples
+
+    def _forward_postprocess(self, num_samples: int, result: EstimatorResult) -> np.ndarray:
+        """Post-processing during forward pass of the network."""
+        if num_samples is None:
+            num_samples = 1
+        expectations = np.reshape(result.values, (-1, num_samples)).T
+        return expectations
+
+    def _forward(
+        self, input_data: np.ndarray | None, weights: np.ndarray | None
+    ) -> np.ndarray | None:
+        """Forward pass of the neural network."""
+        parameter_values_, num_samples = self._preprocess(input_data, weights)
+        if num_samples is None:
+            job = self.estimator.run(self._circuit, self._observables)
+        else:
+            job = self.estimator.run(
+                [self._circuit] * num_samples * self.output_shape[0],
+                [op for op in self._observables for _ in range(num_samples)],
+                np.tile(parameter_values_, (self.output_shape[0], 1)),
+            )
+        try:
+            results = job.result()
+        except Exception as exc:
+            raise QiskitMachineLearningError("Estimator job failed.") from exc
+
+        return self._forward_postprocess(num_samples, results)
+
+    def _backward_postprocess(
+        self, num_samples: int, result: EstimatorGradientResult
+    ) -> tuple[np.ndarray | None, np.ndarray]:
+        """Post-processing during backward pass of the network."""
+        num_observables = self.output_shape[0]
+        if self._input_gradients:
+            input_grad = np.zeros((num_samples, num_observables, self._num_inputs))
+        else:
+            input_grad = None
+
+        weights_grad = np.zeros((num_samples, num_observables, self._num_weights))
+        gradients = np.asarray(result.gradients)
+        for i in range(num_observables):
+            if self._input_gradients:
+                input_grad[:, i, :] = gradients[i * num_samples : (i + 1) * num_samples][
+                    :, : self._num_inputs
+                ]
+                weights_grad[:, i, :] = gradients[i * num_samples : (i + 1) * num_samples][
+                    :, self._num_inputs :
+                ]
+            else:
+                weights_grad[:, i, :] = gradients[i * num_samples : (i + 1) * num_samples]
+        return input_grad, weights_grad
+
+    def _backward(
+        self, input_data: np.ndarray | None, weights: np.ndarray | None
+    ) -> Tuple[np.ndarray | None, np.ndarray]:
+        """Backward pass of the network."""
+        # prepare parameters in the required format
+        parameter_values_, num_samples = self._preprocess(input_data, weights)
+
+        if num_samples is None or not (self._input_gradients or self._num_weights):
```

Comment:
Could this be simpler?

```suggestion
        if num_samples is None or (self._input_gradients is False and self._num_weights == 0):
```

Reply:
`self._input_gradients is False` for None its fine - for False it should be better to do `not self._input_gradients` or ` == False` otherwise `is False` relies on singleton behavior to ensure it `is` the same instance

Reply:
PEP 8: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
So, it is well written. Just double checked, pylint complains on `if xx == False`. And `not` reduces readability in this case. I'm in favor of this style as I suggested this change earlier :)

Reply:
https://peps.python.org/pep-0008/#programming-recommendations

```
Don’t compare boolean values to True or False using ==:

# Correct:
if greeting:

# Wrong:
if greeting == True:

Worse:

# Wrong:
if greeting is True:
```
I thought I had read that some Python implementations do not have singletons hence the `is` would not work in that case. I think your ref was to a flake rule right. Yes == is considered incorrect.